### PR TITLE
Implementation of length penalty using logits processors

### PR DESCRIPTION
**Pull request by maxdebayser**
This implementation is simpler than changing the sampler code and less invasive in the core vllm code.

### CODE REVIEW

File: vllm/entrypoints/grpc/validation.py [hunk 1]
```diff
@@ -74,9 +72,8 @@ def validate_params(params: Parameters, max_max_new_tokens: int):
 
     # Decoding parameter checks
     if decoding.HasField("length_penalty"):
-        # TODO: remove this when we support length penalty
-        TGISValidationError.LengthPenaltyUnsupported.error()
-        if not (1.0 <= decoding.length_penalty.decay_factor <= 10.0):
+        args = [decoding.length_penalty.decay_factor, decoding.length_penalty.decay_factor]
+        if None in args or args[0] < 1.0 or args[1] > 10.0:
```
Comment (njhill) [inline vllm/entrypoints/grpc/validation.py:76]: This doesn't look right to me.. isn't the check that was already there good?

File: vllm/tgis_utils/logits_processors.py [hunk 2]
```diff
@@ -14,3 +14,17 @@ def __call__(self, token_ids: List[int],
         # transformers warpers assume tensors of shape (batch_size, vocab_size)
         # and the typical warper doesn't use input_ids
         return self.warper(input_ids=None, scores=logits.reshape((1, -1)))
+
+
+class LengthPenaltyWarper:
```
Comment (njhill) [inline vllm/tgis_utils/logits_processors.py:19]: Could we use the one from transformers like we're doing with typical_p? https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L1505

File: vllm/entrypoints/grpc/grpc_server.py [hunk 3]
```diff
@@ -309,17 +309,33 @@ async def _validate_and_convert_params(
...
+                    eos_token_id=self.tokenizer.eos_token_id
+                )
+            )
+        
 
         time_limit_millis = stopping.time_limit_millis
         deadline = time.time(
         ) + time_limit_millis / 1000.0 if time_limit_millis > 0 else None
 
+
```
Comment (njhill) [inline vllm/entrypoints/grpc/grpc_server.py:338]: nit: no need for extra blank line

File: vllm/entrypoints/grpc/grpc_server.py [hunk 4]
```diff
@@ -309,17 +309,33 @@ async def _validate_and_convert_params(
 
         # to match TGIS, only including typical_p processing
         # when using sampling
+
+
```
Comment (njhill) [inline vllm/entrypoints/grpc/grpc_server.py:313]: nit: no need for extra blank lines

File: vllm/entrypoints/grpc/grpc_server.py [hunk 5]
```diff
@@ -309,17 +309,33 @@ async def _validate_and_convert_params(
...
+                params.decoding.length_penalty.start_index,
+                params.decoding.length_penalty.decay_factor,
+            )
+            logits_processors.append(
+                LengthPenaltyWarper(
+                    length_penalty=length_penalty,
+                    eos_token_id=self.tokenizer.eos_token_id
+                )
+            )
+
```
Comment (njhill) [inline vllm/entrypoints/grpc/grpc_server.py:332]: nit: no need for extra blank line

**Review by njhill — COMMENTED** [review]
Thanks @maxdebayser! made a few comments inline

File: vllm/tgis_utils/logits_processors.py (hunk 2 above)
Comment (maxdebayser) [reply vllm/tgis_utils/logits_processors.py:19]: Yes, actually I looked at this code but the calculation they do is a bit different than what we do in tgis, so I thought it was more important to replicate the exact behavior we have in tgis, but I can use the huggingface implementation, no problem.
https://github.com/huggingface/transformers/blob/c9f6e5e35156e068b227dd9b15521767f6afd4d2/src/transformers/generation/logits_process.py#L1594
https://github.ibm.com/ai-foundation/fmaas-inference-server/blob/21a6bd62c09878d4c6850428c6a418c4ca2f94e1/server/text_generation_server/utils/tokens.py#L91

File: vllm/entrypoints/grpc/validation.py (hunk 1 above)
Comment (maxdebayser) [reply vllm/entrypoints/grpc/validation.py]: Oh, thanks for spotting that!

File: vllm/tgis_utils/logits_processors.py (hunk 2 above)
Comment (njhill) [reply vllm/tgis_utils/logits_processors.py:19]: Ah ... I think ours was meant to replicate the functionality of what's in transformers (do you think the results of our current TGIS impl would be different to that?) ... probably best to follow transformers anyhow if there is some difference since that was the original intention. Note this is exponential decay length penalty in transformers, not "length_penalty" which is only used for beam search.

Comment (maxdebayser) [reply vllm/tgis_utils/logits_processors.py:19]: If the logits are positive then there's no difference

Comment (maxdebayser) [reply vllm/tgis_utils/logits_processors.py:20]: Done

**Comment by maxdebayser** [comment]
Closing this one because I didn't sign-off all commits. This is fixed in #14
